### PR TITLE
Integrating cadvisor stats to CRI Pod stats collection

### PR DESCRIPTION
**Pull request by abhi**
For kubernetes-incubator/cri-containerd#341

**What this PR does / why we need it**:
This PR contains changes to update kubelet to consume metrics from both CRI and cadvisor for Pod stats. The corresponding cadvisor changes are here https://github.com/google/cadvisor/pull/1795.
Today CRI stats provides only core metrics such as CPU Usage, Memory Usage, Disk Usage. However the stats exposed by kubelet which is a superset of core metrics and Network, User Defined Metrics are not provided by CRI stats. Hence kubelet stats are extracted from 2 sources , CRI stats and cadvisor stats. 
After the change the summary stats for a CRI based runtime (containerd) is as show here:
https://gist.github.com/abhi/d7351861df6430eb4bc5d711d274ec35

This PR also contains test case change to reflect the stats for cri stats provider. 

**Which issue(s) this PR fixes** :
Fixes # https://github.com/kubernetes-incubator/cri-containerd/issues/341

**Special notes for your reviewer**:
https://docs.google.com/presentation/d/1Os3nyMRBlFuiBLCjPgeaPv6jXylrZW5jiDXJejlA3Wg/edit#slide=id.g27cb4cb6d8_0_0

Godep will be updated to main tree after https://github.com/google/cadvisor/pull/1795 PR is merged.

**Release note**:

```release-note
Kubelet can provide full summary api support except container log stats for CRI container runtime now.
```
cc @Random-Liu

### CODE REVIEW

**Comment by dashpole** [comment]
I would prefer to cut and vendor the cadvisor release separately from the implementation PR if possible.  I plan to do this tomorrow, although you are welcome to prepare and review the implementation, and rebase once the cadvisor release lands.

**Comment by Random-Liu** [comment]
@abhi, @dashpole is going to cut a new cadvisor release and vendor it. We could rebase on that one later. I'm going to review the non-Godep part now.

File: pkg/kubelet/stats/cri_stats_provider.go [hunk 1]
```diff
@@ -194,7 +212,10 @@ func (p *criStatsProvider) getFsInfo(storageID *runtimeapi.StorageIdentifier) *c
 	return &fsInfo
 }
 
-func (p *criStatsProvider) makePodStats(podSandbox *runtimeapi.PodSandbox) *statsapi.PodStats {
+func (p *criStatsProvider) makePodStats(
```
Comment (Random-Liu) [inline pkg/kubelet/stats/cri_stats_provider.go]: I prefer keep the original function. Add another function to overwrite the original result.
In this way, it's easier to remove either part in the future.

File: pkg/kubelet/stats/cri_stats_provider.go [hunk 2]
```diff
@@ -242,18 +264,20 @@ func (p *criStatsProvider) makeContainerStats(
...
-		}
+	if caPodStats != nil {
+		result.UserDefinedMetrics = cadvisorInfoToUserDefinedMetrics(caPodStats)
 	}
-	if stats.Memory != nil {
-		result.Memory.Time = metav1.NewTime(time.Unix(0, stats.Memory.Timestamp))
-		if stats.Memory.WorkingSetBytes != nil {
-			result.Memory.WorkingSetBytes = &stats.Memory.WorkingSetBytes.Value
+	if caPodStats.Spec.HasCpu {
+		result.CPU = buildCPUStats(cstat)
```
Comment (Random-Liu) [inline pkg/kubelet/stats/cri_stats_provider.go]: This is not defined...?

**Comment by Random-Liu** [comment]
@abhi Could you squash your 1st and 2nd commits? The 1st commit doesn't compile.

File: pkg/kubelet/stats/cri_stats_provider.go [hunk 3]
```diff
@@ -115,6 +123,12 @@ func (p *criStatsProvider) ListPodStats() ([]statsapi.PodStats, error) {
 			continue
 		}
 
+		caStats, found := caInfos[containerID]
+		if !found {
+			glog.Errorf("Unable to find cadvisor stats for %q", containerID)
+			continue
```
Comment (Random-Liu) [inline pkg/kubelet/stats/cri_stats_provider.go]: We may not want to continue. The goal is that if the container runtime doesn't have cadvisor integration, we only provide CRI stats.

File: pkg/kubelet/stats/cri_stats_provider.go [hunk 4]
```diff
@@ -126,10 +140,15 @@ func (p *criStatsProvider) ListPodStats() ([]statsapi.PodStats, error) {
 		// container belongs to.
 		ps, found := sandboxIDToPodStats[podSandboxID]
 		if !found {
-			ps = p.makePodStats(podSandbox)
+			caPodSandbox, found := caInfos[podSandboxID]
+			if !found {
+				glog.Errorf("Unable to find cadvisor stats for sandbox %q", podSandboxID)
+				continue
```
Comment (Random-Liu) [inline pkg/kubelet/stats/cri_stats_provider.go]: same

File: pkg/kubelet/stats/cri_stats_provider.go (hunk 1 above)
Comment (abhi) [reply pkg/kubelet/stats/cri_stats_provider.go]: done

File: pkg/kubelet/stats/cadvisor_stats_provider.go [hunk 5]
```diff
@@ -74,24 +74,11 @@ func (p *cadvisorStatsProvider) ListPodStats() ([]statsapi.PodStats, error) {
 	if err != nil {
 		return nil, fmt.Errorf("failed to get imageFs info: %v", err)
 	}
-
-	infos, err := p.cadvisor.ContainerInfoV2("/", cadvisorapiv2.RequestOptions{
-		IdType:    cadvisorapiv2.TypeName,
-		Count:     2, // 2 samples are needed to compute "instantaneous" CPU
-		Recursive: true,
-	})
+	infos, err := getCadvisorContainerInfo(p.cadvisor)
```
[the remark this answers is not in the document]
Comment (abhi) [reply pkg/kubelet/stats/cadvisor_stats_provider.go]: I have grouped the common functionality into this. The one in loop will not be needed for cri stats collection. Hence

File: pkg/kubelet/stats/cri_stats_provider.go (hunk 3 above)
Comment (abhi) [reply pkg/kubelet/stats/cri_stats_provider.go]: done

File: pkg/kubelet/stats/cri_stats_provider.go (hunk 4 above)
Comment (abhi) [reply pkg/kubelet/stats/cri_stats_provider.go]: done

File: pkg/kubelet/stats/cri_stats_provider.go [hunk 6]
```diff
@@ -242,18 +265,24 @@ func (p *criStatsProvider) makeContainerStats(
...
 			// populate the two fields here.
 		},
-		// UserDefinedMetrics is not supported by CRI.
 	}
-	if stats.Cpu != nil {
-		result.CPU.Time = metav1.NewTime(time.Unix(0, stats.Cpu.Timestamp))
-		if stats.Cpu.UsageCoreNanoSeconds != nil {
-			result.CPU.UsageCoreNanoSeconds = &stats.Cpu.UsageCoreNanoSeconds.Value
-		}
+	cstat, found := latestContainerStats(caPodStats)
```
[the remark this answers is not in the document]
Comment (abhi) [reply pkg/kubelet/stats/cri_stats_provider.go]: did it similar to how it was for cadvisor stats .

File: pkg/kubelet/stats/cri_stats_provider.go [hunk 7]
```diff
@@ -242,18 +265,24 @@ func (p *criStatsProvider) makeContainerStats(
...
+	if !found {
+		return result
+	}
+	if caPodStats != nil {
+		result.UserDefinedMetrics = cadvisorInfoToUserDefinedMetrics(caPodStats)
```
[the remark this answers is not in the document]
Comment (abhi) [reply pkg/kubelet/stats/cri_stats_provider.go]: done

File: pkg/kubelet/stats/cri_stats_provider.go [hunk 8]
```diff
@@ -242,18 +265,24 @@ func (p *criStatsProvider) makeContainerStats(
...
+	}
+	if caPodStats.Spec.HasCpu {
```
[the remark this answers is not in the document]
Comment (abhi) [reply pkg/kubelet/stats/cri_stats_provider.go]: not required since caPodStats.Spec is not a pointer ?

**Comment by abhi** [comment]
@dashpole thanks. Yes I only added godeps for the PR to be reviewable. I will drop the commit once I rebase on top of your cadvisor PR

**Comment by Random-Liu** [comment]
It seems that you didn't update godep properly. I'll skip the godep change, we'll rebase eventually anyway.

```
I1121 06:37:34.349] [0;31mhack/make-rules/../../hack/verify-godep-licenses.sh[0m
I1121 06:37:34.349] [0;31mhack/make-rules/../../hack/verify-godeps.sh[0m
I1121 06:37:34.350] [0;31mhack/make-rules/../../hack/verify-staging-godeps.sh[0m
```

File: pkg/kubelet/stats/cri_stats_provider.go [hunk 9]
```diff
@@ -131,10 +143,24 @@ func (p *criStatsProvider) ListPodStats() ([]statsapi.PodStats, error) {
 		// container belongs to.
 		ps, found := sandboxIDToPodStats[podSandboxID]
 		if !found {
+			caPodSandbox, found := caInfos[podSandboxID]
```
Comment (Random-Liu) [inline pkg/kubelet/stats/cri_stats_provider.go]: nit: Move this after `makePodStats`.

File: pkg/kubelet/stats/cri_stats_provider.go [hunk 10]
```diff
@@ -131,10 +143,24 @@ func (p *criStatsProvider) ListPodStats() ([]statsapi.PodStats, error) {
...
+			}
 			sandboxIDToPodStats[podSandboxID] = ps
 		}
-		ps.Containers = append(ps.Containers, *p.makeContainerStats(stats, container, &rootFsInfo, uuidToFsInfo))
+		cs := p.makeContainerStats(stats, container, &rootFsInfo, uuidToFsInfo)
+		// If cadvisor stats is available for the container, use it to populate
+		// container stats
+		if caFound {
+			p.addCadvisorContainerStats(cs, &caStats)
+		}
```
Comment (Random-Liu) [inline pkg/kubelet/stats/cri_stats_provider.go]: Should we also report error here? Or should we remove the error above?

File: pkg/kubelet/stats/cri_stats_provider.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Random-Liu) [reply pkg/kubelet/stats/cri_stats_provider.go]: Yeah, you are right. Your logic is similar with previous code. Sorry for my confusing.

**Comment by Random-Liu** [comment]
@derekwaynecarr I approved this milestone yesterday, because I didn't know I need sig TL to approve it. Ping for approve.

File: pkg/kubelet/stats/cri_stats_provider.go (hunk 10 above)
Comment (abhi) [reply pkg/kubelet/stats/cri_stats_provider.go]: There is no error returned from this right ? This only parses the stats and adds the available one's. The only non disrupting error will be if we dont find cadvisor stats which needs to be printed above?

File: pkg/kubelet/stats/cri_stats_provider.go (hunk 9 above)
Comment (abhi) [reply pkg/kubelet/stats/cri_stats_provider.go]: done

File: pkg/kubelet/stats/cri_stats_provider.go (hunk 10 above)
Comment (Random-Liu) [reply pkg/kubelet/stats/cri_stats_provider.go]: I see.

File: pkg/kubelet/stats/cri_stats_provider.go [hunk 11]
```diff
@@ -120,6 +127,11 @@ func (p *criStatsProvider) ListPodStats() ([]statsapi.PodStats, error) {
 			continue
 		}
 
+		caStats, caFound := caInfos[containerID]
```
Comment (Random-Liu) [inline pkg/kubelet/stats/cri_stats_provider.go]: nit: Move this below, it's weird to split the `!caFound` case and `caFound` case.

**Review by Random-Liu — COMMENTED** [review]
Please rebase and address last nit.

**Comment by Random-Liu** [comment]
Will LGTM after rebase.

File: pkg/kubelet/stats/cri_stats_provider.go [hunk 12]
```diff
@@ -336,3 +364,44 @@ func removeTerminatedContainer(containers []*runtimeapi.Container) []*runtimeapi
...
+	}
+}
+
+func getCRICadvisorStats(ca cadvisor.Interface) (map[string]cadvisorapiv2.ContainerInfo, error) {
+	stats := make(map[string]cadvisorapiv2.ContainerInfo)
+	infos, err := getCadvisorContainerInfo(ca)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch cadvisor stats: %v", err)
+	}
+	infos = removeTerminatedContainerInfo(infos)
```
Comment (Random-Liu) [inline pkg/kubelet/stats/cri_stats_provider.go]: Can we move this into `getCadvisorContainerInfo`? Separate them seems error prone.

Comment (abhi) [reply pkg/kubelet/stats/cri_stats_provider.go]: My original change has it. But looks like the pod level stats change will need the fill set of info obtained from cadvisor. https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/stats/cadvisor_stats_provider.go#L95

**Pull request merged**